lib/utils/parse/parse.go

```diff
@@ -237,276 +175,104 @@ func NewAnyMatcher(in []string) (Matcher, error) {
-					return nil, trace.BadParameter("expected 1 argument for %v.%v got %v", namespace, fn, len(n.Args))
-				}
-				result.transform = emailLocalTransformer{}
-				ret, err := walk(n.Args[0], depth+1)
-				if err != nil {
-					return nil, trace.Wrap(err)
-				}
-				result.parts = ret.parts
-				return &result, nil
-			case RegexpNamespace:
-				switch fn {
-				// Both match and not_match parse the same way.
-				case RegexpMatchFnName, RegexpNotMatchFnName:
-					if len(n.Args) != 1 {
-						return nil, trace.BadParameter("expected 1 argument for %v.%v got %v", namespace, fn, len(n.Args))
-					}
-					re, ok := getBasicString(n.Args[0])
-					if !ok {
-						return nil, trace.BadParameter("argument to %v.%v must be a properly quoted string literal", namespace, fn)
-					}
-					var err error
-					result.match, err = newRegexpMatcher(re, false)
-					if err != nil {
-						return nil, trace.Wrap(err)
-					}
-					// If this is not_match, wrap the regexpMatcher to invert it.
-					if fn == RegexpNotMatchFnName {
-						result.match = notMatcher{result.match}
-					}
-					return &result, nil
-				case RegexpReplaceFnName:
-					if len(n.Args) != 3 {
-						return nil, trace.BadParameter("expected 3 arguments for %v.%v got %v", namespace, fn, len(n.Args))
-					}
-					ret, err := walk(n.Args[0], depth+1)
-					if err != nil {
-						return nil, trace.Wrap(err)
-					}
-					result.parts = ret.parts
-					expression, ok := getBasicString(n.Args[1])
-					if !ok {
-						return nil, trace.BadParameter("second argument to %v.%v must be a properly quoted string literal", namespace, fn)
-					}
-					replacement, ok := getBasicString(n.Args[2])
-					if !ok {
-						return nil, trace.BadParameter("third argument to %v.%v must be a properly quoted string literal", namespace, fn)
-					}
-					result.transform, err = newRegexpReplaceTransformer(expression, replacement)
-					if err != nil {
-						return nil, trace.Wrap(err)
-					}
-					return &result, nil
-				default:
-					return nil, trace.BadParameter("unsupported function %v.%v, supported functions are: regexp.match, regexp.not_match", namespace, fn)
-				}
-			default:
-				return nil, trace.BadParameter("unsupported function namespace %v, supported namespaces are %v and %v", call.X, EmailNamespace, RegexpNamespace)
-			}
-		default:
-			return nil, trace.BadParameter("unsupported function %T", n.Fun)
-		}
-	case *ast.IndexExpr:
-		ret, err := walk(n.X, depth+1)
-		if err != nil {
-			return nil, err
-		}
-		result.parts = append(result.parts, ret.parts...)
-		ret, err = walk(n.Index, depth+1)
-		if err != nil {
-			return nil, err
-		}
-		result.parts = append(result.parts, ret.parts...)
-		return &result, nil
-	case *ast.SelectorExpr:
-		ret, err := walk(n.X, depth+1)
-		if err != nil {
-			return nil, err
-		}
-		result.parts = append(result.parts, ret.parts...)
+	expr, ok := result.(Expr)
+	if !ok {
+		panic(fmt.Sprintf("unexpected parser result type %T (this is a bug)", result))
```

Comment:
We should really avoid panicking for any possible user input, is there any way the result could be a literal? I'm thinking of an input like `{{"asdf"}}` or `{{123}}` which would be badly formed but it should be an error rather than a panic (I don't know if those really panic here just wondering if it's possible). Might be better to play it safe and return an error

Reply:
Thanks for catching this. Those two inputs would indeed result in a panic. Fixed in 25048f0a3df35c130a9ec46139d5a7b9f3cee2cd.